src/main/java/org/shanerx/tradeshop/listeners/AdminEventListener.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * NOTICE: All modifications made by others to the source code belong
+ * to the respective contributor. No contributor should be held liable for
+ * any damages of any kind, whether be material or moral, which were
+ * caused by their contribution(s) to the project. See the full License for more information
+ */
+
+package org.shanerx.tradeshop.listeners;
+
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.Sign;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.Action;
+import org.bukkit.event.block.BlockBreakEvent;
+import org.bukkit.event.block.BlockPlaceEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+import org.shanerx.tradeshop.TradeShop;
+import org.shanerx.tradeshop.enumys.Message;
+import org.shanerx.tradeshop.enumys.Permissions;
+import org.shanerx.tradeshop.enumys.ShopType;
+import org.shanerx.tradeshop.objects.Shop;
+import org.shanerx.tradeshop.objects.ShopChest;
+import org.shanerx.tradeshop.utils.Utils;
+
+public class AdminEventListener extends Utils implements Listener {
+
+	private TradeShop plugin;
+
+	public AdminEventListener(TradeShop instance) {
+		plugin = instance;
+	}
+
+	@EventHandler(priority = EventPriority.HIGH)
+	public void onBlockBreak(BlockBreakEvent event) {
+		Player player = event.getPlayer();
+		Block block = event.getBlock();
+
+		if (ShopType.isShop(block)) {
+			Shop shop = Shop.loadShop((Sign) block.getState());
+
+			if (player.hasPermission(Permissions.ADMIN.getPerm()) || event.getPlayer().getUniqueId().equals(shop.getOwner().getUUID())) {
+
+				if (shop.getInventory() != null)
+					shop.getChestAsSC().resetName();
+
+				shop.remove();
+				return;
+			}
+			event.setCancelled(true);
+			player.sendMessage(Message.NO_TS_DESTROY.getPrefixed());
+
+		} else if (plugin.getListManager().isInventory(block.getType())) {
+			if (player.hasPermission(Permissions.ADMIN.getPerm())) {
+				new ShopChest(block.getLocation()).resetName();
+				return;
+			}
+
+			Sign s = findShopSign(block);
+			if (s == null) {
+				new ShopChest(block.getLocation()).resetName();
+				return;
+			}
+
+			if (!ShopType.isShop(s)) {
+				new ShopChest(block.getLocation()).resetName();
+				return;
+			}
+
+			Shop shop = Shop.loadShop(s);
+
+			if (event.getPlayer().getUniqueId().equals(shop.getOwner().getUUID())) {
+				shop.getChestAsSC().resetName();
+				shop.remove();
+				return;
+			}
+
+			event.setCancelled(true);
+			player.sendMessage(Message.NO_TS_DESTROY.getPrefixed());
+		}
+	}
```

Comment:
What exactly happens in the case of Double Chests? Where exactly are the necessary checks performed? ie. looking for the other half to make sure it is not a shop etc.

Reply:
The answer was utils, but I just moved them due to there being a wrapper in ShopChest and them making more sense there.